llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -2809,6 +2809,10 @@ static Instruction *matchFunnelShift(Instruction &Or, InstCombinerImpl &IC,
           match(R, m_And(m_Neg(m_Specific(X)), m_SpecificInt(Mask))))
         return X;
 
+      // (shl ShVal, X) | (lshr ShVal, ((-X) & (Width - 1)))
+      if (match(R, m_And(m_Neg(m_Specific(L)), m_SpecificInt(Mask))))
+        return L;
```

Comment:
Just as valid is `(shl ShVal, X & (Width - 1)) | (lshr ShVal, -X)`. No?
Not that it would be all that common, but could imagine `-X & (Width - 1)` being the simplified part.

Reply:
Emm, I believe this pattern is unlikely to exist in real-world applications :(

Reply:
although just a few lines, seems like little harm. Tend to think unless there is a notable maintenance burden, more coverage is better as there is ALOT of code out there.

Reply:
I am afraid I disagree with you. IMHO we shouldn't fall in the lie of "coverage". I've seen a bunch of useless optimizations introduced into instcombine in recent years. It makes instcombine heavy and time-consuming. 

Therefore, I am not willing to handle more patterns because it doesn't show new optimization opportunities in my benchmark https://github.com/dtcxzyw/llvm-opt-benchmark.

BTW, I plan to remove some dead code in instcombine if I have enough bandwidth :)

cc @nikic Any comments?